jetty-hazelcast/src/main/java/org/eclipse/jetty/hazelcast/session/HazelcastSessionDataStore.java

```diff
@@ -0,0 +1,211 @@
+    @Override
+    public Set<String> doGetExpired( Set<String> candidates )
+    {
+        if (candidates == null || candidates.isEmpty())
+        {
+            return Collections.emptySet();
+        }
+        long now = System.currentTimeMillis();
+        return candidates.stream().filter( candidate -> {
+            if (LOG.isDebugEnabled())
+            {
+                LOG.debug( "Checking expiry for candidate {}", candidate );
+            }
+            try
+            {
+                SessionData sd = load(candidate);
+
+                //if the session no longer exists
+                if (sd == null)
+                {
+                    if (LOG.isDebugEnabled())
+                    {
+                        LOG.debug( "Session {} does not exist in infinispan", candidate );
+                    }
+                    return true;
+                }
+                else
+                {
+                    if (_context.getWorkerName().equals(sd.getLastNode()))
+                    {
+                        //we are its manager, add it to the expired set if it is expired now
+                        if ((sd.getExpiry() > 0 ) && sd.getExpiry() <= now)
+                        {
+                            if (LOG.isDebugEnabled())
+                            {
+                                LOG.debug( "Session {} managed by {} is expired", candidate, _context.getWorkerName() );
+                            }
+                            return true;
+                        }
+                    }
+                    else
+                    {
+                        //if we are not the session's manager, only expire it iff:
+                        // this is our first expiryCheck and the session expired a long time ago
+                        //or
+                        //the session expired at least one graceperiod ago
+                        if (_lastExpiryCheckTime <=0)
+                        {
+                            if ((sd.getExpiry() > 0 ) && sd.getExpiry() < (now - (1000L * (3 * _gracePeriodSec))))
+                            {
+                                return true;
+                            }
+                        }
+                        else
+                        {
+                            if ((sd.getExpiry() > 0 ) && sd.getExpiry() < (now - (1000L * _gracePeriodSec)))
+                            {
+                                return true;
+                            }
+                        }
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                LOG.warn("Error checking if candidate {} is expired so expire it", candidate, e);
+                return true;
+            }
+            return false;
+        } ).collect( Collectors.toSet() );
+    }
+
+    @Override
+    public boolean exists( String id )
+        throws Exception
+    {
+        return this.sessionDataMap.containsKey( id );
```

Comment:
Is there a performance difference in hazelcast between operations over the collection like containsKey() vs get()? Eg I know there is in infinispan and a get() op is actually cheaper.